src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+     : options(std::move(inMap)) 
+{
+
+}
+
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+LoadData(MatType const &images, MatType const &boundaries,\
+     MatType const &segmentations)
+{
+  const size_t num_images = this->options["num_images"];
+  const size_t row_size = this->options["row_size"];
+  const size_t col_size = this->options["col_size"];
+  MatType input_data(num_images * row_size * 5, col_size);
+  // we store the input data as follows: 
+  // images (3), boundaries (1), segmentations (1).
+  size_t loop_iter = num_images * 5;
+  size_t row_idx = 0;
+  size_t col_i = 0, col_s = 0, col_b = 0;
+  for(size_t i = 0; i < loop_iter; ++i)
+  {
+    if (i % 5 == 4)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(segmentations.colptr(col_s),\
+                                  col_size, row_size).t();
+      ++col_s;
+    }
+    else if (i % 5 == 3)
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(boundaries.colptr(col_b),\
+                                  col_size, row_size).t();
+      ++col_b;
+    }
+    else
+    {
+      input_data.submat(row_idx, 0, row_idx + row_size - 1,\
+        col_size - 1) = MatType(images.colptr(col_i),
+                                  col_size, row_size).t();
+      ++col_i;  
+    }
+    row_idx += row_size;
+  }
+  return input_data;
+} 
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+GetFeatureDimension()
+{
+  /*
+  shrink: amount to shrink channels
+  p_size: size of image patches
+  n_cell: number of self similarity cells
+  n_orient: number of orientations per gradient scale
+  */
+  arma::vec FtrDim(2);
+
+  const size_t shrink = this->options["shrink"];
+  const size_t p_size = this->options["p_size"];
+  const size_t n_cell = this->options["n_cell"];
+  const size_t rgbd = this->options["rgbd"];
+  const size_t n_orient = this->options["n_orient"];
+  /*
+  n_color_ch: number of color channels
+  n_grad_ch: number of gradient channels
+  n_ch: total number of channels
+  */
+  size_t n_color_ch;
+  if (this->options["rgbd"] == 0)
+    n_color_ch = 3;
+  else
+    n_color_ch = 4;
+
+  const size_t n_grad_ch = 2 * (1 + n_orient);
+
+  const size_t n_ch = n_color_ch + n_grad_ch;
+  FtrDim[0] = std::pow((p_size / shrink) , 2) * n_ch;
+  FtrDim[1] = std::pow(n_cell , 2) * (std::pow (n_cell, 2) - 1) / 2 * n_ch;
+  return FtrDim;
+}
+
+template<typename MatType, typename CubeType>
+arma::vec StructuredForests<MatType, CubeType>::
+DistanceTransform1D(arma::vec const &f, const size_t n, const double inf)
+{
+  arma::vec d(n), v(n), z(n + 1);
+  size_t k = 0;
+  v[0] = 0.0;
+  z[0] = -inf;
+  z[1] = inf;
+  for (size_t q = 1; q <= n - 1; ++q)
+  {
+    float s  = ( (f[q] + q * q)-( f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    while (s <= z[k])
+    {
+      --k;
```

Comment:
I'm not sure, but if we set k = 0 at the beginning, how can we decrease k, and still get the right index?